File: src/Library/Calculator/AdditiveTerm.cpp
#include "AdditiveTerm.h"
#include "NOperNode.h"
#include "ValueNode.h"
#include "InternalPrinter.h"

#include "Helper.h"

/// <summary>
/// Finds AdditiveTerm to represent node
/// </summary>
AdditiveTerm::AdditiveTerm(const ExpNode &node) {
	InternalPrinter *printer = new InternalPrinter();
	const OperNode *operNode = dynamic_cast<const OperNode*>(&node);
	if (operNode != nullptr && operNode->GetOperator() == Operator::MULTIPLICATION
		&& operNode->GetChild(0).IsNumericalValue()) {
		// Node is a multiply and it's first child is a numerical value
		coefficient_ = operNode->GetChild(0).AsDouble();

		if (operNode->ChildCount() > 2) {
			// Makes new multiply node for non-coefficient terms
			unique_ptr<NOperNode> nOperNode = make_unique<NOperNode>('*');
			for (int i = 1; i < operNode->ChildCount(); i++) {
				nOperNode->AddChild(operNode->GetChild(i).Clone());
			}
			base_ = move(nOperNode);
		}
		else {
			base_ = operNode->GetChild(1).Clone();
		}
		base_string = base_->Print(*printer);
	}
	else {
		coefficient_ = 1;
		base_ = node.Clone();
		base_string = base_->Print(*printer);
	}
	delete printer;
}
/// <summary>
/// Gets Term as an ExpNode
/// </summary>
unique_ptr<ExpNode> AdditiveTerm::AsExpNode() {
	if (coefficient_ == 1) {
		return move(base_);
	}
	else {
		unique_ptr<NOperNode> operNode = make_unique<NOperNode>('*');
		operNode->AddChild(MakeValueNode(coefficient_));
		operNode->AddChild(move(base_));
		return operNode;
	}
}

/// <summary>
/// Adds coefficients
/// </summary>
void AdditiveTerm::AddToCoefficient(AdditiveTerm *other) {
	coefficient_ += other->coefficient_;
}

/// <summary>
/// Checks if two AdditiveTerm have the same base
/// </summary>
bool AdditiveTerm::operator==(const AdditiveTerm &other) {
	return base_string == other.base_string;
}

/// <summary>
/// Compares sort order of this with with
/// </summary>
bool AdditiveTerm::operator<(const AdditiveTerm &other) const {
	if (base_->IsNumericalValue()) {
		return false;
	}
	else if (other.base_->IsNumericalValue()) {
		return true;
	}
	return base_string < other.base_string;
}

File: src/Library/Calculator/ExpressionTree/Terms/AdditiveTerm.cs
// Adam Dernis © 2021

using Calculator.ExpressionTree.Nodes;
using Calculator.ExpressionTree.Nodes.Operators.NOpers;
using Calculator.ExpressionTree.Nodes.Values;
using Calculator.Printers.Default;
using System;

namespace Calculator.ExpressionTree.Terms
{
    public class AdditiveTerm : IEquatable<AdditiveTerm>, IComparable<AdditiveTerm>
    {
        private double _coefficient;
        private ExpNode _base;
        private string _baseString;

        public AdditiveTerm(ExpNode node)
        {
            DefaultPrinter printer = new DefaultPrinter();
            if (node is MultiplicationOperNode mNode && mNode.GetChild(0) is NumericalValueNode nvNode)
            {
                // Node is a multiply and it's first child is a numerical value
                // That numerical value is the coefficient
                _coefficient = nvNode.DoubleValue;

                // Remove coefficient from base
                mNode.RemoveChild(0);
                _base = mNode;
            } else
            {
                _coefficient = 1;
                _base = node;
            }

            _baseString = _base.Print(printer);
        }

        public ExpNode AsExpNode()
        {
            if (_coefficient == 0) return Helpers.MakeNumericalNode(0);
            if (_coefficient == 1) return _base;
            return Helpers.Multiply(_coefficient, _base);
        }

        public void AddToCoefficient(AdditiveTerm other)
        {
            _coefficient += other._coefficient;
        }

        public int CompareTo(AdditiveTerm other)
        {
            if (_base is NumericalValueNode)
            {
                return 1;
            } else if (other._base is NumericalValueNode)
            {
                return -1;
            } else
            {
                return _baseString.CompareTo(other._baseString);
            }
        }

        public bool Equals(AdditiveTerm other)
        {
            return _baseString == other._baseString;
        }
    }
}
